core/src/server/rpc_module.rs

```diff
@@ -754,6 +754,183 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
+	///     });
+	///
+	///     Ok(())
+	/// });
+	/// ```
+	///
+	pub fn register_subscription<R, F>(
+		&mut self,
+		subscribe_method_name: &'static str,
+		notif_method_name: &'static str,
+		unsubscribe_method_name: &'static str,
+		callback: F,
+	) -> Result<&mut MethodCallback, Error>
+	where
+		Context: Send + Sync + 'static,
+		F: (Fn(Params, PendingSubscriptionSink, Arc<Context>) -> R) + Send + Sync + Clone + 'static,
+		R: IntoSubscriptionCloseResponse + Send + 'static,
+	{
+		if subscribe_method_name == unsubscribe_method_name {
+			return Err(Error::SubscriptionNameConflict(subscribe_method_name.into()));
+		}
+
+		self.methods.verify_method_name(subscribe_method_name)?;
+		self.methods.verify_method_name(unsubscribe_method_name)?;
+
+		let ctx = self.ctx.clone();
+		let subscribers = Subscribers::default();
+
+		// Unsubscribe
+		{
+			let subscribers = subscribers.clone();
+			self.methods.mut_callbacks().insert(
+				unsubscribe_method_name,
+				MethodCallback::Unsubscription(Arc::new(move |id, params, conn_id, max_response_size| {
+					let sub_id = match params.one::<RpcSubscriptionId>() {
+						Ok(sub_id) => sub_id,
+						Err(_) => {
+							tracing::warn!(
+								"Unsubscribe call `{}` failed: couldn't parse subscription id={:?} request id={:?}",
+								unsubscribe_method_name,
+								params,
+								id
+							);
+
+							return MethodResponse::response(id, ResponsePayload::result(false), max_response_size);
+						}
+					};
+
+					let key = SubscriptionKey { conn_id, sub_id: sub_id.into_owned() };
+					let result = subscribers.lock().remove(&key).is_some();
+
+					if !result {
+						tracing::debug!(
+							"Unsubscribe call `{}` subscription key={:?} not an active subscription",
+							unsubscribe_method_name,
+							key,
+						);
+					}
+
+					MethodResponse::response(id, ResponsePayload::result(result), max_response_size)
+				})),
+			);
+		}
+
+		// Subscribe
+		let callback = {
+			self.methods.verify_and_insert(
+				subscribe_method_name,
+				MethodCallback::Subscription(Arc::new(move |id, params, method_sink, conn| {
+					let uniq_sub = SubscriptionKey { conn_id: conn.conn_id, sub_id: conn.id_provider.next_id() };
+
+					// response to the subscription call.
+					let (tx, rx) = oneshot::channel();
+					let (accepted_tx, accepted_rx) = oneshot::channel();
+
+					let sub_id = uniq_sub.sub_id.clone();
+					let method = notif_method_name;
+
+					let sink = PendingSubscriptionSink {
+						inner: method_sink.clone(),
+						method: notif_method_name,
+						subscribers: subscribers.clone(),
+						uniq_sub,
+						id: id.clone().into_owned(),
+						subscribe: tx,
+						permit: conn.subscription_permit,
+					};
+
+					// The subscription callback is a future from the subscription
+					// definition and not the as same when the subscription call has been completed.
+					//
+					// This runs until the subscription callback has completed.
+					let res = callback(params, sink, ctx.clone());
```

Comment:
To make sure I get this right, this avoids an allocation because we pass the parameters directly to the callback instead of calling `into_owned`, which can allocate quite a bit since the params might either be frequent or large?